### DISCUSSION TITLE
AutoScroll of Submenus

### DISCUSSION THREAD

**Discussion by Ken-Kauffmann** [arnaud-j4k08/PicoClockGreenEasy#6] [Q&A]
The Autoscroll in Clockui.cpp works by restoring saved index values to m_curFuncIdx, while in the root menu.  The index values are saved during the "Add root level functions" addFunction in the constructor. 

I would like to support an autoscroll while in my HAMClock submenu.  

I created my Submenu with the following statement: 
    Submenu *hamclockSubmenu =
        addFunctionAndReturnPtr<HamclockSubmenu>(&m_rootMenu);  // created as a copy of AlarmSubmenu

I added four subfunctions with the following statements:
    TRACE << "Add functions of the hamclock submenu";  // modelled after alarm menu.  
    hamclockSubmenu->addFunction<Time>(this,Time::HourMin);
    hamclockSubmenu->addFunction<Date>(this,Date::Local);
    hamclockSubmenu->addFunction<Time>(this,Time::UTCHourMin);
    hamclockSubmenu->addFunction<Date>(this,Date::UTC);

As I understand it, the four subfunctions are added to a vector anchored by m_functions, associated with my new submenu.  Indexing into these four entries is relatively trivial, as the first add will be index 0, and the other three will be 1,2,3 respectively. 

The problem that I have not been able to solve on my own is what to test against in the following test for autoscroll.

     if (m_settings.get().autoScroll && 
        m_editedValueIndex == NoEditing && 
	m_currentMenu == &m_rootMenu &&     // if m_currentMenu contains the address of m_rootMenu

While I am in my HAMClock submenu, I believe that m_currentMenu will have the address of m_functions associated with the submenu. 

I had hoped to use the following to pass that address back to clockUi, during the creation of the submenu.

stmt 26 in Submenu.h appears to provide a way of returning that address.  

    std::vector<std::unique_ptr<AbstractFunction>> *menu()
    {
        return &m_functions;
    }

The only place that menu() is used in the entire source appears to be this statement in Clockui.cpp:

    TRACE << "Add functions of the countdown submenu";
    m_countdownFunc = countdownSubmenu->addFunction<Countdown>(this, countdownSubmenu->menu());  // only place that menu() is used - kdk  

I was hoping that I could create a variable called :
int m_HCmenuAddress = 0;        // For comparison against m_currentMenu for auto scroll of Hamclock kdk

and to replace it in the autoscroll test as follows:

    if (m_settings.get().autoScroll && 
        m_editedValueIndex == NoEditing && 
	m_currentMenu == m_HCmenuAddress &&     //Compare against stored address of m_functions for my submenu

I wanted to add the following statement to set m_HCmenuAddress after the creation of the subMenu,

    Submenu *hamclockSubmenu =
        addFunctionAndReturnPtr<HamclockSubmenu>(&m_rootMenu);  // created as a copy of AlarmSubmenu
        m_HCmenuAddress =  hamclockSubmenu->menu();  // Store the address of the submenu m_functions for AutoScroll check

but I don't have the syntax even close to being correct.  

Am I even close?

**Comment by arnaud-j4k08** [comment]
I just checked the code and it looks like you are close to the solution. Maybe the only problem is the type of m_HCmenuAddress, which needs to be a pointer to a vector of unique_ptr to AbstractFunction (we can't put the Function objects directly in a vector, as their actual types differ). 

You can indeed create a submenu function like in the following code. I don't think you need an own HamclockSubmenu class, as you can use Submenu directly. The only reason for the existence of AlarmSubmenu (derived from Submenu) is that its text dynamically mentions the time of the next alarm.
```
    Submenu *hamclockSubmenu= 
        addFunctionAndReturnPtr<Submenu>("Hamclock", &m_rootMenu);
```
A Submenu object is the function that gives access to the submenu. The actual functions of this submenu can indeed be retrieved this way:
`m_hcMenu = hamclockSubmenu->menu();`

Whereas, m_hcMenu is declared as member variable of ClockUi (initialized to nullptr by default to avoid a "random" value):
`std::vector<std::unique_ptr<AbstractFunction>> *m_hcMenu = nullptr;`

Then, you can use it in the block that handles autoscroll. Here is an example that rotates through 4 functions every 15 seconds:
```
        if (m_settings.get().autoScroll && 
            m_editedValueIndex == NoEditing && 
            m_currentMenu == m_hcMenu &&
            m_secondsWithoutUserInput >= AUTO_SCROLL_DELAY_SEC)
        {
            if (m_clock.get().tm_sec % 15 == 0)
            {
                m_curFuncIdx = m_clock.get().tm_sec / 15;
                startVertScrolling(-1);
            }
        }
```

**Comment by Ken-Kauffmann** [comment]
You make it look so easy.  Your changes worked perfectly.  And your block of code for doing the submenu autoscroll used 5 statements, instead of my 22.  Thank you for your help.  The PicoClockGreenEasy is now doing everything that I wanted it to do.
